### ISSUE TITLE
OCaml 4.06 and Map.update

### ISSUE THREAD

**Issue by SkySkimmer** [coq/coq#6196]
In ocaml 4.06 the Map interface provides a function `update : key -> ('a option -> 'a option) -> 'a t -> 'a t` (ocaml/ocaml#1026).
Coq's Map also has a function called update but it is `key -> 'a -> 'a t -> 'a t`, we also have `modify` which is like ocaml's update but without the `option`.

Should we do something about this?

**Comment by psteckler** [comment]
There are about a dozen uses of Map.update, so it wouldn't be a diffcult change to make.

**Comment by ejgallego** [comment]
I think it makes sense to uniformize them indeed.

**Comment by ppedrot** [comment]
+1 for uniformization.

**Comment by psteckler** [comment]
I'm working on this unless someone tells me not to.

**Comment by psteckler** [comment]
How OK is it to start with the OCaml source for this?

The code in `cMap.ml` is (maybe overly) clever, using `Obj.magic` to switch between a locally-declared type and the OCaml representation of maps, so OCaml code works there (modulo some small changes). `update` requires delicate merging and balancing operations, which must be done exactly as OCaml does them to preserve the isomorphism.

**Comment by ppedrot** [comment]
@psteckler I think the solution is simply not to define `update` in `CMap` and simply let the module inclusion take care of that. For this to work with all OCaml version, you should remove the `update` entry from `CMap.ExtS`.

Also, it is probably worthwhile to keep the Coq update under another name. It is indeed more efficient than the OCaml one because it does not need to allocate a closure.

TL;DR: just rename update into something else and this should work out of the box.

**Comment by psteckler** [comment]
Created #6290.
